### PR TITLE
Add initial functionality

### PR DESCRIPTION
**Pull request by dlh01**

### CODE REVIEW

File: src/class-big-pit.php [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+		}
+
+		if ( ! $installed_version ) {
+			\dbDelta(
+				<<<SQL
+CREATE TABLE {$wpdb->big_pit} (
+	item_id bigint(20) UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+	item_group varchar(255) NOT NULL,
+	item_key varchar(255) NOT NULL,
+	item_value longtext NOT NULL
```
Comment (mboynes) [inline src/class-big-pit.php:209]:
```suggestion
	item_value longtext NOT NULL,
	KEY group_key (item_group(191),item_key(191))
```

File: src/class-big-pit.php [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+				'item_group' => $group,
+			],
+			[ '%s' ],
+		);
+	}
+
+	/**
+	 * Create or update the database table.
+	 */
+	private function upsert(): void {
```
Comment (scottnelle) [inline src/class-big-pit.php:188]: Any value in returning `bool` here, `true` if it was already ready or if the db delta completed, `false` if not? You could check that result before declaring the featured `ready`.

File: src/class-big-pit.php [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	}
+
+	/**
+	 * Set a value.
+	 *
+	 * @param string $key   Item key.
+	 * @param mixed  $value Item value.
+	 * @param string $group Item group.
+	 */
+	public function set( string $key, mixed $value, string $group ): void {
```
Comment (scottnelle) [inline src/class-big-pit.php:99]: Might it be beneficial to return `bool` on the `set` and `delete` methods, to mirror the corresponding object cache methods?

**Review by scottnelle — APPROVED** [review]
Couple questions, no blockers. If you'd like to use a more descriptive name for this plugin, don't feel beholden to the name we joked about during retro. But if you want to stick with Big Pit, OK by me! 👻

**Comment by mboynes** [comment]
Wild idea: what if this were fully compatible with the WordPress object cache's (faux) interface? Then it could be an option for a database-driven object cache.

If it _also_ used DI to take in an instance of WPDB instead of using the global `$wpdb`, it could theoretically use a separate database/instance to distribute load (this would make it more attractive as a db object cache, but I think the value is there regardless).

File: src/class-big-pit.php (hunk 3 above)
Comment (mboynes) [reply src/class-big-pit.php:99]: this got me thinking that maybe this should do more than mirror the object cache! expanded on that in another comment

Comment (dlh01) [reply src/class-big-pit.php:104]: I've recently become somewhat skeptical of these return values. In addition to the possibility of ambiguity (does `false` mean the value wasn't deleted or that there was nothing to delete? would you return `true` even if no deletions took place?), have you found a use for them?

**Comment by dlh01** [comment]
> Wild idea: what if this were fully compatible with the WordPress object cache's (faux) interface? Then it could be an option for a database-driven object cache.
> 
> If it _also_ used DI to take in an instance of WPDB instead of using the global `$wpdb`, it could theoretically use a separate database/instance to distribute load (this would make it more attractive as a db object cache, but I think the value is there regardless).

Although I see the harmony, my initial reaction is that I'm not greatly attracted to this idea, for a couple reasons.

Compatibility with the object cache "interface" requires a decent amount of additional code, some of which might be used even when The Pit wasn't being used as the object cache (TTLs), but perhaps more of which would be dead when it wasn't being used as the object cache (hit rates, compatibility with `wp_suspend_cache_addition()`).

Additionally, I think it would compel the table itself to be a global table for multisite compatibility, and keys would need to be made site-specific. Again, that's complexity that runs up against (what I had in mind as) the intended scope of the library and the table.

If there's interest in a database-driven object cache, might I humbly suggest looking at the [Training Object Cache](https://github.com/dlh01/wp-training-object-cache/) that I worked on a few years ago as a starting point, where a lot of that legwork has been done.

See also [the next PR](https://github.com/alleyinteractive/wp-big-pit/pull/2), which opens a possibility for using The Pit as a cache store.

File: src/class-big-pit.php (hunk 1 above)
Comment (dlh01) [reply src/class-big-pit.php:216]: Thank you! I meant to ask you about that.

**Comment by dlh01** [comment]
My plan is to merge this and tag an initial release as there's a project waiting to use it. I'll keep it at major version 0 so that this discussion can continue in issues (including about the name!) and we can refine the API before tagging a 1.0 release.

**Pull request merged**